engines/mus.cpp

```diff
@@ -146,25 +150,48 @@ namespace pono {
 
     /*
      * If the conjunct encodes a state update constraint (i.e. of form
-     * <stateVar>.next = ...), then uses that state variable's id for the
+     * <stateVar>.next = ...), then use that state variable's id for the
      * control var symbol. Otherwise use the conjunct's to_string as an
      * identifier.
      */
-    UnorderedTermMap transIdToConjunct;
+    unordered_map<string, Term> transIdToConjunct;
     for (auto &tc: transConjuncts) {
       Term id = tc;
       if (tc->get_op() == PrimOp::Equal) {
         Term lhs = *tc->begin();
         if (ts_.is_next_var(lhs)) {
-          id = lhs;
+          id = ts_.curr(lhs);
         }
       }
       Term t = unrollUntilBound(tc, k);
       if (!options_.mus_include_yosys_internal_netnames_ && isYosysInternalNetname(id)) {
         solver_->assert_formula(t);
         assertions.push_back(t);
       } else {
-        transIdToConjunct.insert({id, t});
+        transIdToConjunct.insert({id->to_string(), t});
+      }
+    }
+
+    if (!options_.mus_combine_suffix_.empty()) {
+      /*
+       * Conjoin TRANS constraints that are identical up to suffix matching the
+       * supplied regular expression.
+       */
+      std::multimap<string, Term> mm;
+      for (auto & [id, tc] : unordered_map(transIdToConjunct)) {
+        smatch m;
+        regex r(std::regex("(.*)" + options_.mus_combine_suffix_));
+        if (regex_search(id, m, r)) {
+          mm.insert({m[1], tc});
+          transIdToConjunct.erase(id);
+        }
+      }
+      for (auto & [id, tc] : mm) {
+        if (transIdToConjunct.find(id) == transIdToConjunct.end()) {
+          transIdToConjunct[id] = tc;
+        } else {
+          transIdToConjunct[id] = solver_->make_term(And, transIdToConjunct[id], tc);
```

Comment:
When should I expect this path to be exercised?

It looks like it should be exercised when the source program has variables`m`, `m_1`, and `m_2` and that it will then combine the constraints for `m` and (one of) `m_1` or `m_2`. Is that right?

What's the rationale for merging those constraints via `AND` (as opposed to keeping them separate as separate entries in `transIdToConjunct` so that he MUS engine treats them separately)?

Reply:
Almost, see [this comment](https://github.com/GaloisInc/pono/pull/3#discussion_r1662732626) I left above for how it plays out.

> What's the rationale for merging those constraints via AND (as opposed to keeping them separate as separate entries in transIdToConjunct so that he MUS engine treats them separately)?

In the case of `m_1`/`m_2` here, we're assuming they're each a copy of the the constraint for `m` in the source-level program. If the MUS engine were able to consider `m_1`, m_2` each as atomic constraints, then it could yield MUSes with a constraint for just one of the program copies, which would be troublesome to map back to the source-level program.

Reply:
I understand now, thanks.

It is a shame that C++ obfuscates this so much. In Haskell, this would be `map (foldr And True) mm`.